File: code-jam-2021/double-or-noting.py
GLOBAL_MAX = 15

def shiftleft(binary_number):
    return f"{binary_number}0"


def invert(binary_number): 
    result = ""
    started = False

    for bit in binary_number:
        if bit == "0":
            result += "1"
            started = True
        elif bit == "1" and started:
            result += "0"

    return result


def recurse(s, e, current, best):
    if current >= best[0]:
        return 

    if s == e:
        best[0] = min(best[0], current)
        return

    recurse(shiftleft(s), e, current + 1, best)
    recurse(invert(s), e, current + 1, best)

def solve():
    S, E = str(input()).split(" ")

    best = [GLOBAL_MAX]
    recurse(S, E, 0, best)

    if best[0] == GLOBAL_MAX:
        return "IMPOSSIBLE"
    else:
        return best[0]


if __name__ == "__main__":
    T = int(input())
    c = 1

    while T:
        print(f"Case #{c}: {solve()}")
        c += 1
        T -= 1

File: code-jam-2021/double-or-noting.cpp
#include <bits/stdc++.h>

using namespace std;

unordered_set<long long> cache;
int best;

void recurse(unsigned long long & s, unsigned long long & e, int current)
{
    if (s == e)
    {
        if (best == -1 || current < best)
        {
            best = current;
        }
        return;
    }

    if (current >= 5) return;

    s = s << 1;
    cout << "A recursing!: " << s << endl;
    recurse(s, e, current + 1);
    s = s >> 1;
    s = ~s;
    cout << "B recursing!: " << s << endl;
    recurse(s, e, current + 1);
    s = ~s;
}


int solve()
{
    cache = unordered_set<long long>();
    best = -1;

    string S, E;

    cin >> S >> E;

    unsigned long long s = stoi(S, 0, 2);
    unsigned long long e = stoi(E, 0, 2);

    cout << "S: " << s << endl;
    cout << "E: " << e << endl;

    recurse(s, e, 0);
    return best;
}

int main()
{
    int T;
    int c = 1;

    cin >> T;

    while (T--)
    {
        int result = solve();
        
        if (result == -1)
        {
            cout << "Case #" << c << ": IMPOSSIBLE\n";
        }
        else
        {
            cout << "Case #" << c << ": " << result << "\n";
        }

        c++;
    }
}